### PR TITLE
Implement atan2

### PR DESCRIPTION
**Pull request by admcrae**
This addresses #6095.

Since `atan2` is not implemented in Eigen, this is loosely based on the `rint` implementation. Suggestions are welcome.

### CODE REVIEW

File: tensorflow/core/ops/math_ops.cc [hunk 1]
```diff
@@ -732,6 +732,18 @@ The polygamma function is defined as:
 where \\(\psi(x)\\) is the digamma function.
 )doc");
 
+REGISTER_OP("Atan2")
+    .Input("y: T")
```
Comment (drpngx) [inline tensorflow/core/ops/math_ops.cc]: numpy uses x1, x2. It's ugly but we probably want to stay consistent.

File: tensorflow/core/kernels/cwise_ops.h [hunk 2]
```diff
@@ -658,6 +658,22 @@ struct zeta : base<T, Eigen::internal::scalar_zeta_op<T>> {};
 template <typename T>
 struct polygamma : base<T, Eigen::internal::scalar_polygamma_op<T>> {};
 
+template <typename Scalar>
+struct scalar_atan2_op {
+  EIGEN_EMPTY_STRUCT_CTOR(scalar_atan2_op)
+  EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE Scalar
+  operator()(const Scalar& y, const Scalar& x) const {
+#if defined(__CUDACC__)
```
Comment (drpngx) [inline tensorflow/core/kernels/cwise_ops.h]: GOOGLE_CUDA

**Review by drpngx — CHANGES_REQUESTED** [review]

**Comment by drpngx** [comment]
That will require and API review.

**Comment by admcrae** [comment]
I've addressed your code comments (although I won't get a chance to test the GPU code myself until tomorrow).

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Could we add a test that checks random values, and also all special values listed here?
https://docs.scipy.org/doc/numpy/reference/generated/numpy.arctan2.html

Thanks

**Comment by admcrae** [comment]
Done. Gradients for random arguments are rather flaky, unless I clip the inputs to be well away from zero. Perhaps this is because the gradients blow up near the origin. Curiously, I have to clip them more severely for float64 inputs, due to the stricter tolerance on the check.

**Review by drpngx — APPROVED** [review]

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by admcrae** [comment]
Apparently `math.inf` doesn't exist in python 2. I've fixed that, but it doesn't explain the python 3 failure. How can I see the detailed log of the failure? I can't find it on the CI web interface.

**Comment by admcrae** [comment]
FWIW, running bazel test --config=opt //tensorflow/python/kernel_tests:cwise_ops_test passes now (with my previous commit) for both python 2 and 3 on my local machine. Perhaps you could test it again?

**Comment by drpngx** [comment]
You'll need to add an entry [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/api/golden/tensorflow.pbtxt#L635) to extend the public API.

Jenkins, test this please.

**Comment by admcrae** [comment]
Done.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
@martinwicke this is an API change, WDYT?

**Review by martinwicke — CHANGES_REQUESTED** [review]
We looked at this and feel that we should follow the saner convention of y, x. Detailed comments inline.

File: tensorflow/core/ops/math_ops.cc [hunk 3]
```diff
@@ -732,6 +732,17 @@ The polygamma function is defined as:
...
+REGISTER_OP("Atan2")
+    .Input("x1: T")
+    .Input("x2: T")
+    .Output("z: T")
+    .Attr("T: {float, double}")
+    .SetShapeFn(shape_inference::BroadcastBinaryOpShapeFn)
+    .Doc(R"doc(
+Computes arctangent of x1/x2 element-wise, respecting signs of the arguments.
+This is the angle of x2 + x1 * i.
+)doc");
```
Comment (martinwicke) [inline tensorflow/core/ops/math_ops.cc]: After looking at the numpy docs and shivering, we decided it'll be best to stick with the time tested atan2(y,x) convention. Could you revise the docstring and arg names?

For reference: 
Python's math.atan2: https://docs.python.org/2/library/math.html
Numpy's arctan2: https://docs.scipy.org/doc/numpy/reference/generated/numpy.arctan2.html

**Review by martinwicke — CHANGES_REQUESTED** [review]

**Comment by martinwicke** [comment]
This is good from an API perspective after reverting to the `(y, x)` signature.

**Comment by admcrae** [comment]
Done.

**Review by martinwicke — APPROVED** [review]

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by martinwicke** [comment]
Jenkins, test this please.

**Comment by drpngx** [comment]
Fetch errors.

Jenkins, test this please.

**Pull request merged**